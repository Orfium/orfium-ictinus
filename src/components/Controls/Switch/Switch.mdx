import { Meta, Canvas, ArgTypes, Story } from '@storybook/blocks';
import Switch from './Switch';
import * as SwitchStories from './Switch.stories';
import { LabelConfigStory } from '../../storyUtils/LabelConfigStoryComponents';

<Meta of={SwitchStories} />

<SectionHeader title={'Switch'} />

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Variants](#variants)

## Overview

Enables the user to set the state of a single setting to 'on' or 'off'.

## Props

<ArgTypes of={Switch} />

#### LabelConfig Prop

<ArgTypes of={LabelConfigStory} />

## Usage

<UsageGuidelines
  guidelines={[
    'Use switches for binary choices only (e.g. “enable/disable”)',
    'When the item that is being toggled has a default state (e.g. deactivated by default)',
  ]}
  policies={[
    'If you have more than two choices, use a checkbox or a radio button instead',
    'If you can select 1+ items, use checkboxes',
    'If you can select exactly 1 item, use radio buttons',
  ]}
/>

<SubsectionHeader title="Variants" />

### Simple Switch

<Canvas of={SwitchStories.SimpleSwitch} />

### Switch label placement

Label can be placed either `left` or `right` of the switch.

<Tip>
  The labelConfig prop of each Switch includes an sx prop in order to add custom styles on the
  container that wraps the switch input and the label
</Tip>

<Canvas of={SwitchStories.SwitchLabelPlacement} />

### Switch label sizes

Label can be either a string or a custom component. In the case of the string, there are 2 sizes (declared inside the labelConfig), `normal` and `large`

<Canvas of={SwitchStories.SwitchLabelSizes} />

### Switch with helptext

Switch can include a helptext under the label.

<Canvas of={SwitchStories.SwitchWithHelptext} />

### Disabled Switch

<Canvas of={SwitchStories.DisabledSwitch} />

### Playground

<Canvas of={SwitchStories.Playground} />
