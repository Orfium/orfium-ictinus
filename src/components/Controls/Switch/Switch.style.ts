import type { SerializedStyles } from '@emotion/react';
import { css } from '@emotion/react';
import type { Theme } from 'theme';

import { getControlsTokens } from 'components/Controls/Controls.tokens';
import type { LabelConfig } from 'components/Controls/Controls.types';

export const switchStyles =
  ({ placement = 'right', sx }: Pick<LabelConfig, 'placement' | 'sx'>) =>
  (theme: Theme): SerializedStyles => {
    const tokens = getControlsTokens(theme);

    return css`
      display: flex;
      flex-direction: ${placement === 'right' ? 'row' : 'row-reverse'};
      align-items: center;
      gap: ${tokens('switch.padding')};
      position: relative;
      cursor: pointer;

      .bar {
        width: ${tokens('switch.width.track')};
        height: ${tokens('switch.height.track')};
        background: ${tokens('switch.backgroundColor.track')};
        position: absolute;
        border-radius: ${tokens('switch.borderRadius')};
      }

      .indicator {
        width: ${tokens('switch.width.track')};
        height: ${tokens('switch.size.thumb')};
        box-sizing: border-box;
        position: relative;

        &:before {
          content: '';
          box-sizing: border-box;
          display: block;
          width: ${tokens('switch.size.thumb')};
          height: ${tokens('switch.size.thumb')};
          background: ${tokens('switch.backgroundColor.thumb.default')};
          border: ${tokens('switch.borderWidth')} solid
            ${tokens('switch.borderColor.thumb.default')};
          border-radius: 100%;
          transition: all 200ms;
        }
      }

      &[data-hovered],
      &[data-focus-visible='true'] {
        .indicator {
          &:before {
            transition: all 0.2s;
            box-shadow: 0px 0px 0px 8px ${theme.tokens.state.get('backgroundColor.hover')};
            border-radius: 100%;
          }
        }
      }

      &[data-selected] {
        .indicator {
          &:before {
            transform: translateX(80%);
            background: ${tokens('switch.backgroundColor.thumb.active')};
            border-color: ${tokens('switch.borderColor.thumb.active')};
          }
        }
      }

      &[data-disabled] {
        opacity: ${theme.tokens.disabledState.get('default')};
        cursor: not-allowed;
      }

      ${sx};
    `;
  };
