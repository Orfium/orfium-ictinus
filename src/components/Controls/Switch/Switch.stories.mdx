import { Meta, Story, Preview, Props } from '@storybook/addon-docs';
import Switch from './Switch';
import Stack from '../../storyUtils/Stack';
import { FIGMA_URL, Function } from '../../../utils/common';
import { boolean, select, withKnobs, text } from '@storybook/addon-knobs';
import { useState } from 'react';
import Overview from '../../../storybook/Overview';

<Meta title="Updated Components/Controls/Switch" component={Switch} />

<SectionHeader title={'Switch'} />

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Variants](#variants)

## Overview

<Story
  name="Overview"
  parameters={{
    previewTabs: {
      canvas: { hidden: true },
    },
  }}
>
  <Overview>Enables the user to set the state of a single setting to 'on' or 'off'.</Overview>
</Story>

## Props

<Props of={Switch} />

## Usage

<UsageGuidelines
  guidelines={[
    'Use switches for binary choices only (e.g. “enable/disable”)',
    'When the item that is being toggled has a default state (e.g. deactivated by default)',
  ]}
  policies={[
    'If you have more than two choices, use a checkbox or a radio button instead',
    'If you can select 1+ items, use checkboxes',
    'If you can select exactly 1 item, use radio buttons',
  ]}
/>

<SubsectionHeader title="Variants" />

### Simple Switch

<Preview>
  <Story name="Simple Switch">
    <Stack>
      <Function>
        {() => {
          const [selected, setSelected] = useState(false);
          return (
            <Switch isSelected={selected} onChange={setSelected}>
              Option
            </Switch>
          );
        }}
      </Function>
    </Stack>
  </Story>
</Preview>

### Switch label placement

Label can be placed either `left` or `right` of the switch.

<Tip>
  The labelConfig prop of each Switch includes an sx prop in order to add custom styles on the
  container that wraps the switch input and the label
</Tip>

<Preview>
  <Story name="Switch label placement">
    <Stack height={50}>
      <Switch>Option</Switch>
    </Stack>
    <Stack>
      <Switch labelConfig={{ placement: 'left' }}>Option</Switch>
    </Stack>
  </Story>
</Preview>

### Switch label sizes

Label can be either a string or a custom component. In the case of the string, there are 2 sizes (declared inside the labelConfig), `normal` and `large`

<Preview>
  <Story name="Switch label sizes">
    <Stack height={50}>
      <Switch>Normal Option</Switch>
    </Stack>
    <Stack>
      <Switch labelConfig={{ size: 'large' }}>Large Option</Switch>
    </Stack>
  </Story>
</Preview>

### Switch with helptext

Switch can include a helptext under the label.

<Preview>
  <Story name="Switch with helptext">
    <Stack>
      <Switch labelConfig={{ helpText: 'This is the helptext of the option' }}>Option</Switch>
    </Stack>
  </Story>
</Preview>

### Disabled Switch

<Preview>
  <Story name="Disabled Switch">
    <Stack height={50}>
      <Switch isDisabled>Option</Switch>
    </Stack>
    <Stack>
      <Switch labelConfig={{ helpText: 'This option is disabled' }} isDisabled>
        Option
      </Switch>
    </Stack>
  </Story>
</Preview>

### Playground

<Preview>
  <Story name="Playground" parameters={{ decorators: [withKnobs] }}>
    <Stack>
      <Switch
        labelConfig={{
          placement: select('Label placement', ['left', 'right'], 'right'),
          size: select('Label size', ['normal', 'large'], 'normal'),
          helpText: text('Help text', ''),
        }}
        isDisabled={boolean('isDisabled', false)}
      >
        Option
      </Switch>
    </Stack>
  </Story>
</Preview>
