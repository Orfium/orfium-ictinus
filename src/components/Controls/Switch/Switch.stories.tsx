import { FIGMA_URL } from 'utils/common';
import Switch from './Switch';
import { useState } from 'react';
import Stack from 'components/storyUtils/Stack';
import { boolean, select, text } from '@storybook/addon-knobs';

export default {
  title: 'Updated Components/Controls/Switch',
  component: Switch,
  parameters: {
    design: [
      {
        type: 'figma',
        name: 'Switch',
        url: `${FIGMA_URL}?node-id=10283%3A104364`,
      },
    ],
  },
};

export const SimpleSwitch = {
  render: () => {
    const [selected, setSelected] = useState(false);
    return (
      <Switch isSelected={selected} onChange={setSelected}>
        Option
      </Switch>
    );
  },
  name: 'Simple Switch',
};

export const SwitchLabelPlacement = {
  render: () => {
    return (
      <>
        <Stack height={50}>
          <Switch>Option</Switch>
        </Stack>
        <Stack>
          <Switch labelConfig={{ placement: 'left' }}>Option</Switch>
        </Stack>
      </>
    );
  },
  name: 'Switch label placement',
};

export const SwitchLabelSizes = {
  render: () => {
    return (
      <>
        <Stack height={50}>
          <Switch>Normal Option</Switch>
        </Stack>
        <Stack>
          <Switch labelConfig={{ size: 'large' }}>Large Option</Switch>
        </Stack>
      </>
    );
  },
  name: 'Switch label sizes',
};

export const SwitchWithHelptext = {
  render: () => {
    return <Switch labelConfig={{ helpText: 'This is the helptext of the option' }}>Option</Switch>;
  },
  name: 'Switch with helptext',
};

export const DisabledSwitch = {
  render: () => {
    return (
      <>
        <Stack height={50}>
          <Switch isDisabled>Option</Switch>
        </Stack>
        <Stack>
          <Switch labelConfig={{ helpText: 'This option is disabled' }} isDisabled>
            Option
          </Switch>
        </Stack>
      </>
    );
  },
  name: 'Disabled Switch',
};

export const Playground = {
  render: () => {
    return (
      <Stack>
        <Switch
          labelConfig={{
            placement: select('Label placement', ['left', 'right'], 'right'),
            size: select('Label size', ['normal', 'large'], 'normal'),
            helpText: text('Help text', ''),
          }}
          isDisabled={boolean('isDisabled', false)}
        >
          Option
        </Switch>
      </Stack>
    );
  },
  name: 'Playground',
};
