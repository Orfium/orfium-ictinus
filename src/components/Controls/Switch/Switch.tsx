import * as React from 'react';
import type { SwitchAria } from 'react-aria';
import { Switch as ReactAriaSwitch } from 'react-aria-components';
import type { TestProps } from 'utils/types';

import { switchStyles } from './Switch.style';
import ControlLabel from 'components/Controls/ControlLabel';
import type { LabelConfig } from 'components/Controls/Controls.types';

export type SwitchProps = Partial<SwitchAria> & {
  /** Id property of the switch element */
  id?: string;
  /** The value of the switch element */
  value?: string;
  /** Callback for when the element's selection state changes. */
  onChange?: (isSelected: boolean) => void;
  /** Label configuration; includes placement, size, helpText and sx */
  labelConfig?: LabelConfig;
  children?: React.ReactNode;
} & TestProps;

const Switch = React.forwardRef<HTMLInputElement, SwitchProps>((props, ref) => {
  const {
    id,
    value,
    isSelected,
    isDisabled,
    onChange,
    labelConfig = {},
    dataTestPrefixId,
    children,
  } = props;
  const { placement = 'right', size = 'normal', helpText, sx } = labelConfig;

  return (
    <ReactAriaSwitch
      id={id}
      value={value}
      isSelected={isSelected}
      isDisabled={isDisabled}
      onChange={onChange}
      css={switchStyles({ placement, sx })}
      data-testid={`${dataTestPrefixId}_${value}_switch`}
      ref={ref}
    >
      <div className="bar" />
      <div className="indicator" />
      {children && (
        <ControlLabel
          size={size}
          helpText={helpText}
          dataTestPrefixId={`${dataTestPrefixId}_radio_${value?.split(' ').join('_')}`}
        >
          {children}
        </ControlLabel>
      )}
    </ReactAriaSwitch>
  );
});

Switch.displayName = 'Switch';

export default Switch;
