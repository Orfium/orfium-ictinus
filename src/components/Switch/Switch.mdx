import { Meta, Canvas, ArgTypes } from '@storybook/blocks';
import Switch from './Switch';
import Stack from '../storyUtils/Stack';
import SwitchShowcase from '../storyUtils/SwitchShowcase';
import { boolean, select, withKnobs, text } from '@storybook/addon-knobs';
import * as SwitchStories from './Switch.stories';

<Meta of={SwitchStories} />

# Switch

An interactive Switch component meant to serve as an alternative way of toggling between states.

## Usage

```js
import { Switch } from '@orfium/ictinus';

<Switch />;
```

## Props

<ArgTypes of={Switch} />

### Simple Switch

<Canvas of={SwitchStories.SimpleSwitch} />
