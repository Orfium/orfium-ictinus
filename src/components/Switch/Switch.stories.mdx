import { Meta, Story, Preview, Props } from '@storybook/addon-docs';
import Switch from './Switch';
import Stack from '../storyUtils/Stack'; import SwitchShowcase from "../storyUtils/SwitchShowcase"; import {
    boolean,
    select, withKnobs
} from "@storybook/addon-knobs";

<Meta title="Design System/Switch" component={Switch} />

# Switch

An interactive Switch component meant to serve as an alternative way of toggling between states.

# Usage

```js
import { Switch } from '@orfium/ictinus';

<Switch />
```

# Props

<Props of={Switch} />

# Simple Switch

<Preview>
  <Story name="Simple Switch" parameters={{ decorators: [withKnobs] }}>
    <Stack>
      <SwitchShowcase
          hasLabel={boolean('hasLabel', true)}
          labelPlacement={select('labelPlacement', ['left', 'right'], 'left')} />
    </Stack>
  </Story>
</Preview>

