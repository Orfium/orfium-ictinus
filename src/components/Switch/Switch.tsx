import { rgba } from 'polished';
import React from 'react';
import ReactSwitch from 'react-switch';

import { useTheme } from '../../index';
import { TestProps } from '../../utils/types';
import { Label, SwitchWrapper, Container } from './Switch.style';

interface Props {
  label?: string;
  labelPlacement?: 'left' | 'right';
  checked: boolean;
  onChange: (
    checked?: boolean,
    event?: React.SyntheticEvent<MouseEvent | KeyboardEvent> | MouseEvent
  ) => void;
  disabled?: boolean;
}

const Switch: React.FC<Props & TestProps> = ({
  disabled = false,
  label,
  checked,
  onChange,
  labelPlacement = 'left',
  dataTestId = 'switch',
}) => {
  const theme = useTheme();

  const addLabel = (component: JSX.Element) => {
    if (!label) {
      return component;
    }

    return (
      <label>
        <Container>
          {labelPlacement === 'left' ? (
            <>
              <Label data-testid={`${dataTestId}_label`}>{label}</Label>
              {component}
            </>
          ) : (
            <>
              {component}
              <Label>{label}</Label>
            </>
          )}
        </Container>
      </label>
    );
  };

  return addLabel(
    <SwitchWrapper checked={checked} disabled={disabled}>
      <ReactSwitch
        data-testid={dataTestId}
        checked={checked}
        onChange={onChange}
        offHandleColor={theme.utils.getColor('lightGrey', 50)}
        offColor={theme.utils.getColor('lightGrey', 150)}
        onHandleColor={theme.utils.getColor('blue', 500)}
        onColor={theme.utils.getColor('blue', 150)}
        handleDiameter={20}
        uncheckedIcon={false}
        checkedIcon={false}
        height={8}
        width={30}
        activeBoxShadow={`${rgba(14, 14, 23, 0.1)} 0 0 2px 3px`}
        disabled={disabled}
      />
    </SwitchWrapper>
  );
};

Switch.displayName = 'Switch';

export default Switch;
