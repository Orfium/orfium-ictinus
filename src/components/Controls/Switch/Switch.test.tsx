import userEvent from '@testing-library/user-event';
import React from 'react';

import { render, screen } from '../../../test';
import Switch from './Switch';
import { Mock } from 'vitest';

describe('Switch', () => {
  let mockOnClick: Mock<any, any>;

  beforeEach(() => {
    mockOnClick = vi.fn();
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  it('it renders the Switch correctly', () => {
    const { container } = render(<Switch>Label</Switch>);

    expect(container).toMatchSnapshot();
  });

  it('should be able to change its check condition', async () => {
    render(<Switch />);

    const switchComponent = screen.getByTestId('ictinus_switch');

    expect(switchComponent.getAttribute('data-selected')).toEqual(null);

    await userEvent.click(switchComponent);

    expect(switchComponent.getAttribute('data-selected')).toEqual('true');
  });

  it('should invoke the onChange function', async () => {
    render(<Switch isSelected={false} onChange={mockOnClick} />);
    const switchComponent = screen.getByTestId('ictinus_switch');

    await userEvent.click(switchComponent);

    expect(mockOnClick).toHaveBeenCalledTimes(1);
  });

  it('should not invoke the onChange function if the switch is disabled', async () => {
    render(<Switch isSelected={false} onChange={mockOnClick} isDisabled />);
    const switchComponent = screen.getByTestId('ictinus_switch');

    await userEvent.click(switchComponent);

    expect(mockOnClick).toHaveBeenCalledTimes(0);
  });
});