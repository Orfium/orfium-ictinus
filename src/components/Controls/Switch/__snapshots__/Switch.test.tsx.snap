// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Switch > it renders the Switch correctly 1`] = `
.emotion-0 [data-disabled] {
  opacity: 50%;
  cursor: not-allowed;
}

.emotion-1 {
  cursor: pointer;
}

.emotion-1>div>div:first-of-type {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 0.75rem;
  position: relative;
}

.emotion-1 .bar {
  width: 2.25rem;
  height: 0.5rem;
  background: #B8CCFA;
  position: absolute;
  border-radius: 624.9375rem;
}

.emotion-1 .indicator {
  width: 2.25rem;
  height: 1.25rem;
  box-sizing: border-box;
  position: relative;
}

.emotion-1 .indicator:before {
  content: '';
  box-sizing: border-box;
  display: block;
  width: 1.25rem;
  height: 1.25rem;
  background: #ffffff;
  border: 0.125rem solid #8EAAEC;
  border-radius: 100%;
  -webkit-transition: all 200ms;
  transition: all 200ms;
}

.emotion-1[data-hovered] .indicator:before,
.emotion-1[data-focus-visible='true'] .indicator:before {
  -webkit-transition: all 0.2s;
  transition: all 0.2s;
  box-shadow: 0px 0px 0px 8px rgba(200,206,255,0.20);
  border-radius: 100%;
}

.emotion-1[data-selected] .indicator:before {
  -webkit-transform: translateX(80%);
  -moz-transform: translateX(80%);
  -ms-transform: translateX(80%);
  transform: translateX(80%);
  background: #175BF5;
  border-color: #175BF5;
}

.emotion-2 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.75rem;
  -webkit-box-pack: unset;
  -ms-flex-pack: unset;
  -webkit-justify-content: unset;
  justify-content: unset;
}

.emotion-6 {
  font-family: Roboto;
  font-weight: 400;
  line-height: 1.25rem;
  font-size: 0.875rem;
  letter-spacing: 0.015625rem;
  color: #111530;
}

<div>
  <div
    class="emotion-0"
  >
<<<<<<< HEAD
    <label
=======
    <span
      style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
    >
      <input
        role="switch"
        type="checkbox"
      />
    </span>
    <div
      class="bar"
    />
    <div
      class="indicator"
    />
    <div
>>>>>>> 4d208708 (feat: fix all snapshots and tests)
      class="emotion-1"
      data-rac=""
      data-testid="ictinus_switch"
    >
      <span
        style="border: 0px; clip-path: inset(50%); height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px; white-space: nowrap;"
      >
        <input
          role="switch"
          tabindex="0"
          type="checkbox"
        />
      </span>
      <div
        class="emotion-2 emotion-3"
      >
        <div
          class="emotion-4 emotion-3"
        >
          <div
            class="bar"
          />
          <div
            class="indicator"
          />
        </div>
        <div
          class="emotion-6"
          data-testid="ictinus_switch_undefined_label"
        >
          Label
        </div>
      </div>
    </label>
  </div>
</div>
`;
